src/consensus/blockchain.py

```diff
@@ -515,6 +508,81 @@ async def pre_validate_blocks_multiprocessing(
         self,
         blocks: List[FullBlock],
     ) -> Optional[List[PreValidationResult]]:
-        return await pre_validate_blocks_multiprocessing(
-            self.constants, self.constants_json, self.sub_blocks, self.sub_height_to_hash, blocks, self.pool
+        return await pre_validate_blocks_multiprocessing(self.constants, self.constants_json, self, blocks, self.pool)
+
+    def sub_block_record(self, header_hash: bytes32) -> SubBlockRecord:
+        return self.__sub_blocks[header_hash]
+
+    def height_to_sub_block_record(self, sub_height: uint32) -> SubBlockRecord:
+        header_hash = self.sub_height_to_hash(sub_height)
+        return self.sub_block_record(header_hash)
+
+    def get_ses_heights(self) -> List[uint32]:
+        return sorted(self.__sub_epoch_summaries.keys())
+
+    def get_ses(self, height: uint32) -> SubEpochSummary:
+        return self.__sub_epoch_summaries[height]
+
+    def sub_height_to_hash(self, height: uint32) -> Optional[bytes32]:
+        return self.__sub_height_to_hash[height]
+
+    def contains_sub_height(self, height: uint32) -> bool:
+        return height in self.__sub_height_to_hash
+
+    def get_peak_height(self) -> Optional[uint32]:
+        return self.peak_height
+
+    async def warmup(self, fork_point: uint32):
+        if self.peak_height is None:
+            return
+        blocks = await self.block_store.get_sub_block_in_range(
+            max(fork_point - self.constants.SUB_BLOCKS_CACHE_SIZE, 0), self.peak_height
         )
+        self.__sub_blocks = blocks
+        return
+
+    def clean_sub_block_record(self, sub_height: int):
+        if sub_height < 0:
+            return
+        blocks_to_remove = self.__sub_heights_in_cache.get(uint32(sub_height), None)
+        while blocks_to_remove is not None and sub_height >= 0:
+            log.debug(f"delete sub height {sub_height} from sub blocks")
+            for header_hash in blocks_to_remove:
+                del self.__sub_blocks[header_hash]  # remove from sub blocks
+            del self.__sub_heights_in_cache[uint32(sub_height)]  # remove height from heights in cache
+
+            sub_height = sub_height - 1
+            blocks_to_remove = self.__sub_heights_in_cache.get(uint32(sub_height), None)
+
+    def clean_sub_block_records(self):
+        if len(self.__sub_blocks) < self.constants.SUB_BLOCKS_CACHE_SIZE:
+            return
+
+        peak = self.get_peak()
+        assert peak is not None
+        if peak.sub_block_height - self.constants.SUB_BLOCKS_CACHE_SIZE < 0:
+            return
+        self.clean_sub_block_record(peak.sub_block_height - self.constants.SUB_BLOCKS_CACHE_SIZE)
+
+    async def get_sub_blocks_in_range(self, start: int, stop: int) -> Dict[bytes32, SubBlockRecord]:
```

Comment:
Where is this method used?

Reply:
weight proofs __create_sub_epoch_segments, wouldn't need it if there was an easy way to check if a header block was a challenge block without needing the prev sub blocks to check deficit.